File: 2018/python/day02.py
import itertools
import math
from collections import Counter

from aocd import get_data, submit

DAY = 2
YEAR = 2018


def main() -> tuple[int, str]:
    data = get_data(day=DAY, year=YEAR).split("\n")
    freqs = (Counter(word) for word in data)
    part1 = math.prod(
        map(sum, zip(*((2 in freq.values(), 3 in freq.values()) for freq in freqs)))
    )

    part2 = ""
    for w1, w2 in itertools.combinations(data, 2):
        diffs = 0
        diff_pos = 0
        for c1, c2, idx in zip(w1, w2, range(len(w1))):
            if c1 == c2:
                continue
            diffs += 1
            diff_pos = idx
            if diffs > 1:
                break
        if diffs == 1:
            part2 = "".join(c for i, c in enumerate(w1) if i != diff_pos)
            break

    return part1, part2


if __name__ == "__main__":
    ans1, ans2 = main()
    submit(ans1, part="a", day=DAY, year=YEAR)
    submit(ans2, part="b", day=DAY, year=YEAR)

File: 2018/rust/src/bin/day02.rs
use aocd::*;
use itertools::Itertools;
use std::error::Error;

#[aocd(2018, 2)]
pub fn main() {
    let input: Vec<String> = input!().lines().map(|line| line.to_string()).collect();
    submit!(1, part_1(&input));
    submit!(2, part_2(&input).unwrap());
}

pub fn part_1(input: &Vec<String>) -> isize {
    let mut two = 0;
    let mut three = 0;
    for word in input {
        let freq = word.chars().fold([0u8; 26], |mut counter, c| {
            counter[c as usize - 'a' as usize] += 1;
            counter
        });
        if freq.contains(&2) {
            two += 1
        };
        if freq.contains(&3) {
            three += 1
        };
    }
    two * three
}

pub fn part_2(input: &Vec<String>) -> Result<String, Box<dyn Error>> {
    for (w1, w2) in input.iter().tuple_combinations() {
        let mut diff = 0;
        let mut idx = 0;
        for (i, (c1, c2)) in w1.chars().zip(w2.chars()).enumerate() {
            if c1 == c2 {
                continue;
            }
            diff += 1;
            if diff > 1 {
                break;
            }
            idx = i;
        }
        if diff == 1 {
            let mut result = w1.clone();
            result.replace_range(idx..idx + 1, "");
            return Ok(result);
        }
    }
    panic!("No solution found")
}
